### ISSUE TITLE
It shows and runs in the lock screen

### ISSUE THREAD

**Issue by conualfy** [micheleg/dash-to-dock#726]
I think this is more likely a Gnome Shell bug, but I should also write it here. When the screen is locked, nothing should be available. But I see the bar and I can start applications from it, even though I cannot see them running, but when I unlock, I can see them. This is a huge security risk and should be fixed asap in Gnome.
![gnome-shell-security-bug-20180427_224542](https://user-images.githubusercontent.com/475091/39382175-5bd82186-4a6d-11e8-9a40-7b9607b7b073.jpg)

**Comment by micheleg** [comment]
Hi, this is the common behaviour when "anything" goes wrong. The extension fails to disable itself when you lock the screen. This is typically due to some incompatibility with other extensions. I recommend you to 

1. Install the latest version (v63). It's still not accepted on the extension website so you'll have to install it manually, e.g. https://micheleg.github.io/dash-to-dock/releases.html

2. If this doesn't help, It is known that on ubuntu uninstalling ubuntu-dock might fix the issue, although this is more a workaround than a clean solution. You might want to try this.

**Comment by MilyMilo** [comment]
Hey, I can say this in fact exist even with dash-to-dock v63. 
It's quite tricky, since it only appears from time to time, either there's a dock in the login screen or I have a duplicated (vanilla settings) dock on the left while my one is also visible on the bottom.
A temporary fix is to restart gnome-shell with `alt+f2` then `r` and enter.

**Comment by irakliy01** [comment]
I have the same issue. Moreover, after rebooting or locking screen I see two bars. As I remember, this bug was in 17.10 too.
[Screenshot](https://imgur.com/a/GyBPkpi)

**Comment by conualfy** [comment]
@micheleg  I run version 63 already. The Ubuntu one is not active due to some incompatibility. Anyways, my idea is that Gnome should never allow anything but the safe things on the lock screen, the simplest way is to cover with a locked layer that covers everything else and blocks creating some other windows on top of that. I have reported it to Gnome, but they said it is the extension's problem, not Gnome's fault and closed my report. :(
https://gitlab.gnome.org/GNOME/gnome-shell/issues/237

**Comment by conualfy** [comment]
@micheleg  I run version 63 already. The Ubuntu one is not active due to some incompatibility. Anyways, my idea is that Gnome should never allow anything but the safe things on the lock screen, the simplest way is to cover with a locked layer that covers everything else and blocks creating some other windows on top of that. I have reported it to Gnome, but they said it is the extension's problem, not Gnome's fault and closed my report. :(
https://gitlab.gnome.org/GNOME/gnome-shell/issues/237

**Comment by ywwg** [comment]
I see this problem in ubuntu 18.04 too.  
I can't uninstall the ubuntu dock without removing the ubuntu desktop metapackage.  (Mine is also disabled anyway)

**Comment by SMadani** [comment]
This happens to me every time on Ubuntu 18.04

**Comment by d2a-raudenaerde** [comment]
Still happens. Just upgraded from 16.04 to 18.04.

**Comment by Mbuntu** [comment]
@micheleg Youre first method ("Install the latest version (v63). It's still not accepted on the extension website so you'll have to install it manually, e.g. https://micheleg.github.io/dash-to-dock/releases.html") is worked for me. Thank you very much. :)

**Comment by micheleg** [comment]
@Mbuntu v63 should be available on the extension website, what do you mean by "It's still not accepted "?

**Comment by osilkin98** [comment]
I'm still having this issue; on multiple devices. Maybe @micheleg  should do a better job of looking into issue tickets.

**Comment by conualfy** [comment]
I still experience the bug, too. Never saw any difference in this domain. Just logged out, moved the cursor to see the locked screen and started Terminal from my left bar.

**Comment by ProfElectric** [comment]
All of these comments seem to have Ubuntu in common.  Indeed, when I'm running Fedora, I don't have this issue, but Ubuntu 17.10+ does produce the issue reliably.

It's a hack and it's not clean, but I have found the following works: 

1. Install Dash to Dock
2. Disable Dash to Dock
3. Disable Ubuntu Dock (may already be disabled)
4. sudo apt remove gnome-shell-extension-ubuntu-dock (this will remove ubuntu-desktop metapackage as well)
5. Re-enable Dash to Dock
6. Enjoy until the next upgrade.

The issue seems to be specifically how Ubuntu is deploying their dock.

**Comment by conualfy** [comment]
> All of these comments seem to have Ubuntu in common. Indeed, when I'm running Fedora, I don't have this issue, but Ubuntu 17.10+ does produce the issue reliably.
> 
> It's a hack and it's not clean, but I have found the following works:
> 
>     1. Install Dash to Dock
> 
>     2. Disable Dash to Dock
> 
>     3. Disable Ubuntu Dock (may already be disabled)
> 
>     4. sudo apt remove gnome-shell-extension-ubuntu-dock (this will remove ubuntu-desktop metapackage as well)
> 
>     5. Re-enable Dash to Dock
> 
>     6. Enjoy until the next upgrade.
> 
> 
> The issue seems to be specifically how Ubuntu is deploying their dock.

How did you do that? I just tried and it wants to uninstall Ubuntu Desktop.

```
The following packages will be REMOVED:
  gnome-shell-extension-ubuntu-dock ubuntu-desktop

```

**Comment by osilkin98** [comment]
@ProfElectric Yeah, I believe I mentioned that in one of my issues, it's definitely an issue with Canonical's fork of Gnome. I've experimented with normal Gnome and didn't really have an issue.

**Comment by osilkin98** [comment]
@conualfy 
> Later edit: you already said that. I am not sure what Ubuntu Desktop is, but if it is the Desktop, I want it, so cannot remove the package

It's not the desktop but rather a list of packages that it states ***must*** be installed as its dependencies. Since the default `gnome-shell-extension-ubuntu-dock` is one of the packages within the dependencies, asking apt to remove it will also cause it to remove the **list** of dependencies that must be installed with `ubuntu-desktop`, but **not the** ***software*** in the list. 
In other words, if you think of the `ubuntu-desktop` as being a condition that is satisfied when all the packages within its list are installed, if you remove a package within the list, the condition is no longer true, and so `apt` removes it as well.

 There's a lot of [documentation](https://help.ubuntu.com/community/MetaPackages) about meta-packages in Ubuntu, so hopefully this will clear things up.

**Comment by brianc118** [comment]
Experiencing this as well on a fresh 18.04 install

**Comment by Ads20000** [comment]
This seems to be a duplicate of #649 and #641. The Ubuntu bug is [here](https://bugs.launchpad.net/ubuntu/+source/gnome-shell-extension-ubuntu-dock/+bug/1769383), please mark yourself as affected by it and pay attention to when developers ask for more information so that they can fix the bug (I've provided the journal log when the bug occurs)!

**Comment by mike10004** [comment]
I noted the following over at the [Launchpad report](https://bugs.launchpad.net/ubuntu/+source/gnome-shell-extension-ubuntu-dock/+bug/1769383): Open `/usr/share/gnome-shell/extensions/ubuntu-dock@ubuntu.com/extension.js` and somewhere around line 31, change

    dockManager.destroy()

to

    if (dockManager != null) {
        dockManager.destroy();
    }

and the dock will stop appearing on the lock screen.

**Comment by mike10004** [comment]
If I understand what's going on correctly, `dockManager` is declared globally, so it might still be an issue that Ubuntu Dock and Dash To Dock each construct a dock manager instance, but `destroy()` only gets called on one of them. A complete resolution would need each extension to retain a reference to the instance it creates.

**Comment by pdrd** [comment]
@mike10004 I understand your thoughts, but i took a look at the source code of dash-to-dock and gnome-shell-extension-ubuntu-dock and the `dockManager` variables in the extension.js files are file-local and conditionally set to `new Docking.DockManager();`.

What may be the problem is that in the [gnome-shell-extension-ubuntu-dock/tree/extension.js](https://git.launchpad.net/ubuntu/+source/gnome-shell-extension-ubuntu-dock/tree/extension.js?h=ubuntu%2Fbionic) `enable()` function the `dockManager` is initialized with `null` and the scenario i can imagine is that the `enable()` and `disable()` extension hooks are getting called by the extension system even if the ubuntu dock extension is not visible, cause otherwise your fix would take no effect. It seems that in our case the condition to set the local `dockManager` instance evaluates to false and the `disable()` call throws a typical uncatched _TypeError_.

@rugk I think it should be reported to the gnome-shell-extension-ubuntu-dock issue tracker and should not be addressed as dash-to-dock issue. The referenced issue #829 contains an error stack trace, which should not be affected by the seems-to-work-fix of @mike10004, so we might want to separate the issues?

**Comment by rugk** [comment]
I can actually reproduce this issue on non-Ubuntu distro(s), i.e. Fedora. See https://github.com/micheleg/dash-to-dock/issues/829. As such it's likely not really related to Ubuntu – or there are multiple causes for that issue.

**Comment by pdrd** [comment]
That's quite interesting, because i'm working since two days on a ubuntu workstation with the fix, where i'm repeatedly locking/unlocking and the issue didn't came back since then - beforehand it was easily reproduceable. Might be multiple causes..

**Comment by furqaaaan** [comment]
I found a workaround for this that works for me. After customizing however i want using dash to dock, simply disable it. The settings will persist. No more having two docks or docks appearing at lockscreen

**Comment by micheleg** [comment]
Closing as duplicate  of #641.
